contracts/src/types/ProofRequest.sol

```diff
@@ -64,32 +65,54 @@ library ProofRequestLibrary {
         );
     }
 
-    /// @notice Verifies the client's signature over the proof request.
+    /// @notice Verifies a prover signature over a proof request.
+    /// @dev Supports EIP-1271 for smart contract signatures.
     /// @param structHash The EIP-712 struct hash of the proof request.
-    /// @param addr The address of the client.
+    /// @param addr The address of the prover.
     /// @param signature The signature to validate.
-    /// @return The struct hash if the signature is valid.
-    function verifyClientSignature(ProofRequest calldata, bytes32 structHash, address addr, bytes calldata signature)
+    function verifyProverSignature(ProofRequest calldata, bytes32 structHash, address addr, bytes calldata signature)
         internal
-        pure
-        returns (bytes32)
+        view
     {
-        if (ECDSA.recover(structHash, signature) != addr) {
-            revert IBoundlessMarket.InvalidSignature();
+        if (addr.code.length == 0) {
+            // Standard EOA flow.
+            if (ECDSA.recover(structHash, signature) != addr) {
+                revert IBoundlessMarket.InvalidSignature();
+            }
+        } else {
+            // If the signature is not empty and its a smart contract client, we call isValidSignature with the provided signature.
+            // This is the standard flow for a client using a smart contract wallet to submit requests.
+            if (IERC1271(addr).isValidSignature(structHash, signature) != IERC1271.isValidSignature.selector) {
+                revert IBoundlessMarket.InvalidSignature();
+            }
         }
-        return structHash;
     }
 
-    /// @notice Extracts the prover's signature for the given proof request.
+    /// @notice Verifies a signature over a proof request. Additionally validates the flag provided by the client.
+    /// @dev Supports EIP-1271 for smart contract signatures.
     /// @param structHash The EIP-712 struct hash of the proof request.
-    /// @param proverSignature The prover's signature to extract.
-    /// @return The address of the prover.
-    function extractProverSignature(ProofRequest calldata, bytes32 structHash, bytes calldata proverSignature)
-        internal
-        pure
-        returns (address)
-    {
-        return ECDSA.recover(structHash, proverSignature);
+    /// @param addr The address of the client.
+    /// @param signature The signature to validate.
+    /// @param isSmartContractSig Whether the signature should be validated as a smart contract signature.
+    function verifyClientSignature(
+        ProofRequest calldata,
+        bytes32 structHash,
+        address addr,
+        bytes calldata signature,
+        bool isSmartContractSig
+    ) internal view {
```

Comment:
From a cryptography API perspective, this is problematic. We shouldn't pass the hash into this function, as the hash is a core part of the signature check. If the hash is not calculated correctly, the signature is unsound. This is why before we passed the hash as a return arg to still allow reuse of the calculation. Additionally, passing in `isSmartContractSig` is less likely to be unsafe, but is kind of odd. Checking that flag should only take a small amount of gas, so even if it ends up repeated, that seems ok.

Reply:
I think this should be addressed. Unfortunately I had to move the logic out of the ProofRequest library as I can't access the OZ hashing functions which are intended to enable upgradability of the domain (unclear if we'd ever use this, but I thought best to leave support for it for now).